threestudio/models/renderers/nerf_volume_renderer.py

```diff
@@ -176,6 +176,27 @@ def forward(
                     **geo_out,
                 }
             )
+            if "normal" in geo_out:
+                # if self.material.cfg.lambda_normal_smooth > 0: # TODO
+                comp_normal: Float[Tensor, "Nr 3"] = nerfacc.accumulate_along_rays(
+                    weights[..., 0],
+                    values=geo_out["normal"],
+                    ray_indices=ray_indices,
+                    n_rays=n_rays,
+                )
+                comp_normal = F.normalize(comp_normal, dim=-1)
+                comp_normal = (comp_normal + 1.0) / 2.0 * opacity  # for visualization
+                out.update(
+                    {
+                        "comp_normal": comp_normal.view(batch_size, height, width, 3),
+                    }
+                )
+                # if self.material.cfg.lambda_3d_normal_smooth > 0: # TODO
+                normal_perturb = self.geometry(
```

Comment:
consider adding an option to control to perturb or not since this will also introduce extra useless inference for other methods

Reply:
I think a temporal solution could be adding options to the renderer. `self.cfg.return_comp_normal` and `self.cfg.return_normal_perturb` sound fine to me.

Reply:
I reused `lambda_normal_smooth` and `lambda_3d_normal_smooth` from the yaml file